### PR TITLE
Web backend for Jupyter rich-display integration

### PR DESCRIPTION
**Pull request by plbalbi**
In order to display a model with the Jupyter rich-display mechanism, we need to add a _repr_html_ method to the Model class. The main idea is to use the DEVSDiagrammer project, take whats necessary, and integrate it into Jupyter.

It seems simple at first, because it's just a static HTML, but two important things are requirements in order to pull this out:
- We need to embed the JSON representation of the model inside the HTML template. For this, we use tornado templates.
- We need some way to serve all the static files DEVSDiagrammer needs to cosume (css styles, js scripts, and js vendor files). To do this, an approach similar to what matplotlib does was used (https://github.com/matplotlib/matplotlib/blob/d49c7c063065cceaa02b20c8b7c947adb7180df7/lib/matplotlib/backends/backend_webagg.py). 

## How the second item was handled?
In summary, in consisted on starting a separate thread the first time the _repr_html_ is called, and serving the static content with a tornado-web-server.
Some tricky things had to be handled in a hacky way:
- The way the thread starts up, in which we need the _repr_html_ call to block until the web-server starts up.
- The tornado ioloop shutdown, which it appears it's always tricky.

### CODE REVIEW

File: pringles/backends/web_display.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        cls.url_prefix = url_prefix
+
+        if port is None:
+            # TODO: Add random free-port lookup
+            port = 10982
+
+        # Add logic to grab a random *available* port
+        app.listen(port)
+        cls.initialized = True
+        cls.started_latch.release()
```
Comment (plbalbi) [inline pringles/backends/web_display.py]: Need to add here some logic that when the initialization fails, the started latch has to be released anyways

File: pringles/backends/web_display.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    @classmethod
+    def initialize(cls, url_prefix: str = '', port: int = None, address: str = None):
+        if cls.initialized:
+            return
+
+        app = cls(url_prefix=url_prefix)
+        cls.url_prefix = url_prefix
+
+        if port is None:
+            # TODO: Add random free-port lookup
```
Comment (plbalbi) [inline pringles/backends/web_display.py]: This is a MUST

Comment (plbalbi) [reply pringles/backends/web_display.py]: Matplotlib has a nice and simple way of doing this. Check it out: https://github.com/matplotlib/matplotlib/blob/d49c7c063065cceaa02b20c8b7c947adb7180df7/lib/matplotlib/backends/backend_webagg.py#L209

File: tests/test_model_ipython_display.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import pytest
+import requests
+from pringles.models import Coupled
+from .utils import make_queue_top_model_with_events
+
+
+@pytest.fixture(autouse=True)
+def _start_web_display_backend():
+    _start_server()
```
Comment (plbalbi) [inline tests/test_model_ipython_display.py]: In here, mock the `_get_static_files_path` to move test static files to another folder.

File: tests/utils.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+from typing import List, Tuple
+import pytest  # noqa
+from pringles.utils import VirtualTime
+from pringles.models import Coupled, Model, AtomicModelBuilder, Event
+
+
+def make_queue_top_model_with_events() -> Tuple[Model, List[Event]]:
```
Comment (plbalbi) [inline tests/utils.py]: Change all calls to this method in other test suites to use this

File: setup.py [hunk 5]
```diff
@@ -22,5 +22,6 @@
         "License :: OSI Approved :: GNU General Public License v3 (GPLv3)",
         "Operating System :: OS Independent",
     ],
-    install_requires=dependencies
+    install_requires=dependencies,
+    include_package_data=True,
```
Comment (juanlanuzag) [inline setup.py]: Is this to include the html, js, css?

File: pringles/backends/web_display.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+def _get_static_files_path() -> str:
+    # TODO: Overwrite this function on test-suite to move test statics
+    return os.path.join(
+        os.path.dirname(__file__), 'statics')
+
+
+def ipython_inline_display(model: Model) -> bytes:
+    import tornado.template
+    from pringles.serializers import JsonSerializer
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: why do you import these things here and not in the first lines?

File: pringles/backends/web_display.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+    if not web_model_display_thread.is_alive():
+        web_model_display_thread.start()
+        WebApplication.started_latch.acquire()
+
+    single_model_template = Path(_get_static_files_path(), 'basic.html').read_bytes()
+    single_template = tornado.template.Template(single_model_template)
+    return single_template.generate(
+        model_source=JsonSerializer.serialize(model),
+        url_prefix="http://localhost:10982"
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: This shouldn't be hardcoded like this

File: pringles/backends/statics/basic.html [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+<html class="basic-page">
+
+<head>
+  <meta charset="UTF-8">
+  <title>DEVS Diagrammer</title>
+  <meta name="viewport" content="width=device-width, user-scalable=no">
+
+  <link rel="stylesheet" href="{{ url_prefix }}/_static/css/style.css" />
+
+  <script src="{{ url_prefix }}/_static/vendors/js/createjs/createjs-2013.12.12.min.js"></script>
```
Comment (juanlanuzag) [inline pringles/backends/statics/basic.html]: Related to other comment about hardcoded "localhost:<aPort>":

Doesn't it work if you just put /_static/... without the prefix? I think if you do that it resolves the url as being from the same domain

File: pringles/backends/statics/basic.html [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+<body>
+
+  <!-- hiden button for export JSON -->
+  <a id="download" style="display:none"></a>
+  <!-- hiden button for export JSON -->
+
+  <div id="devs_diagrammer">
+    <div id='body-wrapper'>
+      <div id="models-wrapper"></div>
+      <!-- <div id="menu-wrapper">
```
Comment (juanlanuzag) [inline pringles/backends/statics/basic.html]: why delete the buttons?

File: pringles/backends/web_display.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    url_prefix = ''
+    # Hacky way to force the ipython render method to wait for the tornado server to start
+    started_latch = threading.Semaphore(0)
+
+    # Display model as html request handler
+    class DisplayModel(tornado.web.RequestHandler):
+        def get(self):
+            self.write("Not implemented yet")
+
+    class TestGetWithBodyHandler(tornado.web.RequestHandler):
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: This shouldn't be here

File: pringles/backends/web_display.py [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+        def get(self):
+            written_response = "holis"
+            self.write(written_response)
+
+    def __init__(self, url_prefix: str = ''):
+        super().__init__([
+            (url_prefix + r'/test', self.TestGetWithBodyHandler),
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: This shouldn't be here

File: pringles/backends/web_display.py [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        ioloop.start()
+        ioloop.close()
+
+    # Overriding join method to mark the ioloop to be disposed
+    def join(self):
+        ServerThread._join_called = True
+        super().join()
+
+
+web_model_display_thread = ServerThread()
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: maybe move this down closer to the ipython_inline_display method ? Here it's just lost between two classes

File: pringles/backends/web_display.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+class WebApplication(tornado.web.Application):
+
+    initialized = False
+    url_prefix = ''
+    # Hacky way to force the ipython render method to wait for the tornado server to start
+    started_latch = threading.Semaphore(0)
+
+    # Display model as html request handler
+    class DisplayModel(tornado.web.RequestHandler):
```
Comment (juanlanuzag) [inline pringles/backends/web_display.py]: I don't really get what this is for...

File: MANIFEST.in [hunk 14]
```diff
@@ -0,0 +1 @@
+recursive-include pringles/backends/statics *
```
Comment (juanlanuzag) [inline MANIFEST.in]: whats this??

**Comment by juanlanuzag** [comment]
> This pull request **introduces 1 alert** when merging [884b2f2](https://github.com/colonelpringles/pringles/commit/884b2f2588d3e26b368324b7fb67856d89382948) into [ab82ad6](https://github.com/colonelpringles/pringles/commit/ab82ad6659cfc43184910689b9faebf14fd0ab75) - [view on LGTM.com](https://lgtm.com/projects/g/colonelpringles/pringles/rev/pr-f617103ecbe4f9bd7302c19eeacec2ffe1633e8d)
> 
> **new alerts:**
> 
> * 1 for Module is imported with 'import' and 'import from'

@plbalbi check this thing too.

**Comment by juanlanuzag** [comment]
We should also delete the display module (and add the display options to this implementation)

File: setup.py (hunk 5 above)
Comment (plbalbi) [reply setup.py]: Yes, it includes everything in the `backends/statics` path. This is defined by the MANIFEST.in file.

File: pringles/backends/web_display.py (hunk 6 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: The serializer and tornado.templates are just used inside this method. Through reading more implementations, I recon it's a good patter to import things that are needed in just one method from there.

File: pringles/backends/web_display.py (hunk 7 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: Changed with the random port algorithm

File: pringles/backends/statics/basic.html (hunk 8 above)
Comment (plbalbi) [reply pringles/backends/statics/basic.html]: The prefix is needed. The web-server that serve this statics is listening a port different from the one jupyter is binded on. So if you don't add the`{{ url_prefix }}`, it will try to fetch each static from the jupyter URL.

File: pringles/backends/statics/basic.html (hunk 9 above)
Comment (plbalbi) [reply pringles/backends/statics/basic.html]: The whole UI is messed up when inserting it into jupyter. For now, I'll leave them this way. 
Once you check the whole CSS config, you can uncomment them.

File: pringles/backends/web_display.py (hunk 10 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: It's a healthcheck handler. I just renamed it.

File: pringles/backends/web_display.py (hunk 11 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: Yes. That need to be there. 
When you start a tornado web app, you need to call the parent constructor, which is responsible for registering each handler you want.

File: pringles/backends/web_display.py (hunk 12 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: Agree.

File: pringles/backends/web_display.py (hunk 13 above)
Comment (plbalbi) [reply pringles/backends/web_display.py]: Erased. Created before discovering tornado.templates.

File: MANIFEST.in (hunk 14 above)
Comment (plbalbi) [reply MANIFEST.in]: See https://github.com/colonelpringles/pringles/pull/55#discussion_r322025149

**Comment by juanlanuzag** [comment]
> This pull request **introduces 1 alert** when merging [884b2f2](https://github.com/colonelpringles/pringles/commit/884b2f2588d3e26b368324b7fb67856d89382948) into [ab82ad6](https://github.com/colonelpringles/pringles/commit/ab82ad6659cfc43184910689b9faebf14fd0ab75) - [view on LGTM.com](https://lgtm.com/projects/g/colonelpringles/pringles/rev/pr-f617103ecbe4f9bd7302c19eeacec2ffe1633e8d)
> 
> **new alerts:**
> 
> * 1 for Module is imported with 'import' and 'import from'

And fix this
